Source/CarthageKit/Xcode.swift

```diff
@@ -1227,29 +1257,11 @@ public func buildInDirectory(directoryURL: NSURL, withConfiguration configuratio
 			// `.NoSharedFrameworkSchemes`.
 			.filter { projects in !projects.isEmpty }
 			.flatMap(.Merge) { (projects: [(ProjectLocator, [String])]) -> SignalProducer<(String, ProjectLocator), CarthageError> in
-				return SignalProducer(values: projects)
-					.map { (project: ProjectLocator, schemes: [String]) in
-						// Only look for schemes that actually reside in the project
-						let containedSchemes = schemes.filter { (scheme: String) -> Bool in
-							if let schemePath = project.fileURL.URLByAppendingPathComponent("xcshareddata/xcschemes/\(scheme).xcscheme").path {
-								return NSFileManager.defaultManager().fileExistsAtPath(schemePath)
-							}
-							return false
-						}
-						return (project, containedSchemes)
-					}
-					.filter { (project: ProjectLocator, schemes: [String]) in
-						switch project {
-						case .ProjectFile where !schemes.isEmpty:
-							return true
-
-						default:
-							return false
-						}
-					}
-					.concat(SignalProducer(error: .NoSharedFrameworkSchemes(.Git(GitURL(directoryURL.path!)), platforms)))
-					.take(1)
-					.flatMap(.Merge) { project, schemes in SignalProducer(values: schemes.map { ($0, project) }) }
+				let frameworkSchemes = frameworkSchemesForProjects(projects)
+				if frameworkSchemes.last() == nil {
+					return SignalProducer(error: .NoSharedFrameworkSchemes(.Git(GitURL(directoryURL.path!)), platforms))
+				}
+				return frameworkSchemes
```

Comment:
Instead of using `last`, making this function block and do work eagerly, it would be better to do it like so:

```swift
return frameworkSchemesForProjects(projects)
    .collect()
    .flatMap(.Merge) { schemes in
        if schemes.count > 0 {
            return SignalProducer(values: schemes)
        } else {
            return SignalProducer(error: .NoSharedFrameworkSchemes(.Git(GitURL(directoryURL.path!)), platforms))
        }
    }
```

(I haven't tried to compile that code, but it should at least give you the rough idea.)